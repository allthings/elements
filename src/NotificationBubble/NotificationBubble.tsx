import * as React from 'react'
import View from '../View'
import { css } from 'glamor'
import Icon from '../Icon'
import Text from '../Text'
import { Motion, spring } from 'react-motion'
import Theme from '../Theme'

const styles = {
  bubble: css({
    zIndex: 200,
    position: 'fixed',
    bottom: 50,
    left: '50%',
    transform: 'translate(-50%, 0)',
    margin: '0 auto',
    width: 260,
    height: 68,
    borderRadius: 45,
    boxShadow: '2px 2px 14px 1px rgba(35,35,35,0.2)',
  }),
  text: css({
    width: 200,
    marginLeft: 20,
  }),
}

interface INotificationBubble {
  readonly color?: string
  readonly onTimeout?: () => void
}
interface INotificationBubbleState {
  visible: boolean
}

/**
 * NotificationBubbles are used to inform users about important
 * information, when an action has failed or succeeded. They're a most
 * prominent ways to communicate to merchants. NotificationBubbles are
 * always visible and pop up at the bottom of a page.
 *
 * ## Best practises
 * - Do use them when you otherwise users won't have any feedback.
 * - Don't use them extensive.
 * - Don't use them when no feedback is required.
 *
 * To use NotificationBubbles you need to wrap the **NotificationBubbleManager** around your app.
 * Use `sendSuccess()` exported by the NotificationBubbleManager to show a Notification.
 *
 * ```example
 * import NotificationBubbleManager, { sendSuccess } from '@allthings/elements/behaviour/NotificationBubbleManager'
 * import { Button } from '@allthings/elements'
 *
 * const ShowNotification = () => (
 *   <NotificationBubbleManager>
 *     <View>
 *       <Button onClick={() => sendSuccess('Congratulations, you clicked the Button')}>
 *         Click me!
 *       </Button>
 *     </View>
 *   </NotificationBubbleManager>
 * )
 * ```
 *
 */
class NotificationBubble extends React.Component<
  INotificationBubble,
  INotificationBubbleState
> {
  state = {
    visible: true,
  }

  static defaultProps = {
    color: 'primary',
  }

  componentDidMount() {
    setTimeout(() => this.setState({ visible: false }), 2750)
  }

  handleRest = () =>
    this.state.visible === false &&
    this.props.onTimeout &&
    this.props.onTimeout()

  render() {
    const { color = 'primary', onTimeout, ...props } = this.props

    return (
      <Theme>
        {({ colorize }) => (
          <Motion
            defaultStyle={{ position: -50 }}
            onRest={this.handleRest}
            style={{
              position: spring(this.state.visible ? 50 : -100, {
                stiffness: 180,
                damping: 12,
              }),
            }}
          >
            {style => (
              <View {...props}>
                <View
                  {...styles.bubble}
                  style={{
                    backgroundColor: colorize(color),
                    bottom: style.position,
                  }}
                  direction="row"
                  alignV="center"
                >
                  <View flex={70} direction="row" {...styles.text}>
                    <Text
                      color="textOnBackground"
                      align="center"
                      autoBreak
                      style={{ width: '100%' }}
                    >
                      {this.props.children}
                    </Text>
                  </View>
                  <View
                    flex={30}
                    alignH="center"
                    alignV="center"
                    direction="row"
                  >
                    <Icon name="check-filled" color="white" />
                  </View>
                </View>
              </View>
            )}
          </Motion>
        )}
      </Theme>
    )
  }
}

export default NotificationBubble
