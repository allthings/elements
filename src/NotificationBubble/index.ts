export { default } from './NotificationBubble'
export * from './NotificationBubble'
