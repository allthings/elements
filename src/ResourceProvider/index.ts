export { default } from './ResourceProvider'
export * from './ResourceProvider'
