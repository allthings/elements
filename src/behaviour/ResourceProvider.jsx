import React from 'react'
import PropTypes from 'prop-types'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResouceProvider to let the components know where they are.
 */
class ResourceProvider extends React.Component {
  getChildContext() {
    return { resourcePath: this.props.resourcePath }
  }

  static childContextTypes = {
    resourcePath: PropTypes.string,
  }

  static propTypes = {
    resourcePath: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  }

  static defaultProps = {
    resourcePath: RESOURCE_PATH,
    children: null,
  }

  render() {
    return this.props.children
  }
}

export default ResourceProvider
