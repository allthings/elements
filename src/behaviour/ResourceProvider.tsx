import React from 'react'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResouceProvider to let the components know where they are.
 */
const ResouceProvider = React.createContext({
  resourcePath: RESOURCE_PATH,
})

export type ResourceType = {
  resourcePath?: string
}

export default ResouceProvider
