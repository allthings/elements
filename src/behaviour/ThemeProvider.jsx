import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast, Subscriber } from 'react-broadcast'
import { ColorPalette } from '@allthings/colors'

export const THEME_CHANNEL = 'theme'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

/**
* All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
* The ThemeProvider allows you to define the default colors for most elements.
*
* **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
**/
export default class ThemeProvider extends React.PureComponent {
  static propTypes = {
    root: PropTypes.bool,
    theme: PropTypes.shape({
      primary: PropTypes.string,
      text: PropTypes.string,
      secondaryText: PropTypes.string,
      titleColor: PropTypes.string,
      contrast: PropTypes.string,
      warn: PropTypes.string,
      disabled: PropTypes.string,
      background: PropTypes.string,
      textOnBackground: PropTypes.string,
    }),
    children: PropTypes.node,
  }

  static defaultProps = {
    theme: defaultTheme,
  }

  static contextTypes = Subscriber.contextTypes

  render() {
    let root = true
    if (this.context.broadcasts && this.context.broadcasts[THEME_CHANNEL]) {
      root = false
    }
    const { theme, children } = this.props
    return root ? (
      <Broadcast
        channel={THEME_CHANNEL}
        value={{
          ...ThemeProvider.defaultProps.theme,
          ...theme,
        }}
      >
        {children}
      </Broadcast>
    ) : (
      <Subscriber channel={THEME_CHANNEL}>
        {previousTheme => (
          <Broadcast
            channel={THEME_CHANNEL}
            value={{
              ...ThemeProvider.defaultProps.theme,
              ...previousTheme,
              ...theme,
            }}
          >
            {children}
          </Broadcast>
        )}
      </Subscriber>
    )
  }
}

export const ThemeShape = ThemeProvider.propTypes.theme

export const withTheme = (mapThemeToProps, displayName) => WrappedComponent =>
  // eslint-disable-next-line
  class extends React.PureComponent {
    static displayName = displayName || WrappedComponent.displayName
    static component = WrappedComponent

    renderComponent = theme => {
      const props = !mapThemeToProps
        ? { theme }
        : mapThemeToProps(theme, this.props)

      return <WrappedComponent {...this.props} {...props} />
    }

    render() {
      return (
        <Subscriber channel={THEME_CHANNEL}>{this.renderComponent}</Subscriber>
      )
    }
  }
