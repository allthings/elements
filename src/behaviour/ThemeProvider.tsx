import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  primary: string
  text: string
  secondaryText: string
  titleColor: string
  contrast: string
  warn: string
  disabled: string
  background: string
  textOnBackground: string
  [key: string]: string
}

const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = {},
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider

type mapThemeToPropsType = (theme: ITheme, props: any) => any

function createThemeConsumer<T>(
  mapThemeToProps: mapThemeToPropsType,
  WrappedComponent: React.ComponentType<T>,
  props: Omit<T, 'theme'>
) {
  return (theme: ITheme) => {
    const extraProps = !mapThemeToProps
      ? { theme }
      : mapThemeToProps(theme, props)

    return <WrappedComponent {...props} {...extraProps} />
  }
}

export function withTheme(mapThemeToProps: mapThemeToPropsType) {
  return function<T extends { theme: ITheme }>(
    WrappedComponent: React.ComponentType<T>
  ): React.ComponentType<Omit<T, 'theme'>> {
    return function(props: Omit<T, 'theme'>) {
      return (
        <ThemeConsumer>
          {createThemeConsumer<T>(mapThemeToProps, WrappedComponent, props)}
        </ThemeConsumer>
      )
    }
  }
}
