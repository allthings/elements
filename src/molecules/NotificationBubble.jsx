import React from 'react'
import View from '../atoms/View'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'
import Icon from '../atoms/Icon'
import Text from '../atoms/Text'
import { Motion, spring } from 'react-motion'
import PropTypes from 'prop-types'

const styles = {
  bubble: css({
    zIndex: 200,
    position: 'fixed',
    bottom: 50,
    left: '50%',
    transform: 'translate(-50%, 0)',
    margin: '0 auto',
    width: 260,
    height: 68,
    borderRadius: 45,
    boxShadow: '2px 2px 14px 1px rgba(35,35,35,0.2)',
  }),
  text: css({
    width: 200,
    marginLeft: 20,
  }),
}

class NotificationBubble extends React.Component {
  static propTypes = {
    color: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
    onTimeout: PropTypes.func.isRequired,
  }

  state = {
    visible: true,
  }

  static defaultProps = {
    color: 'primary',
    onTimeout: _ => _,
  }

  componentDidMount() {
    setTimeout(() => this.setState({ visible: false }), 2000)
  }

  handleRest = () => this.state.visible === false && this.props.onTimeout()

  render() {
    // eslint-disable-next-line no-unused-vars
    const { color, ...props } = this.props
    return (
      <Theme>
        {({ colorize }) => (
          <Motion
            defaultStyle={{ position: -50 }}
            onRest={this.handleRest}
            style={{
              position: spring(this.state.visible ? 50 : -100, {
                stiffness: 180,
                damping: 12,
              }),
            }}
          >
            {style => (
              <View {...styles.container}>
                <View
                  {...styles.bubble}
                  style={{backgroundColor: colorize(color), bottom: style.position}}
                  direction="row"
                  alignV="center"
                >
                  <View flex={70} direction="row" {...styles.text}>
                    <Text
                      color="textOnBackground"
                      align="center"
                      autoBreak
                      style={{ width: '100%' }}
                    >
                      {this.props.children}
                    </Text>
                  </View>
                  <View
                    flex={30}
                    alignH="center"
                    alignV="center"
                    direction="row"
                  >
                    <Icon name="check-filled" color="white" />
                  </View>
                </View>
              </View>
            )}
          </Motion>
        )}
      </Theme>
    )
  }
}

export default NotificationBubble
