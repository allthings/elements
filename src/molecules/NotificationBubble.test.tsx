import React from 'react'
import renderer from 'react-test-renderer'
import NotificationBubble from './NotificationBubble'
import ThemeProvider from '../behaviour/ThemeProvider'
import ResourceProvider from '../behaviour/ResourceProvider'

test('NotificationBubble renders without error', () => {
  const tree = renderer
    .create(
      <ResourceProvider resourcePath="https://example.com/static/">
        <ThemeProvider>
          <NotificationBubble>Hello World</NotificationBubble>
        </ThemeProvider>
      </ResourceProvider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
