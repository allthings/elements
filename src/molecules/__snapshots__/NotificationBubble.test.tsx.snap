// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBubble renders without error 1`] = `
.css-0,
[data-css-0] {
  z-index: 200;
  position: fixed;
  bottom: 50px;
  left: 50%;
  transform: translate(-50%, 0);
  margin: 0 auto;
  width: 260px;
  height: 68px;
  border-radius: 45px;
  box-shadow: 2px 2px 14px 1px rgba(35,35,35,0.2);
  -webkit-transform: translate(-50%, 0);
}

.css-1,
[data-css-1] {
  width: 200px;
  margin-left: 20px;
}

.css-2,
[data-css-2] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-3,
[data-css-3] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-4,
[data-css-4] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 1 1 70%;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 1 1 70%;
}

.css-5,
[data-css-5] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  text-align: center;
  white-space: pre-wrap;
  word-break: break-word;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #ffffff;
}

.css-6,
[data-css-6] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 1 1 30%;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 1 1 30%;
}

.css-7,
[data-css-7] {
  width: 27.5px;
  height: 27.5px;
  fill: white;
}

<div
  data-css-2=""
>
  <div
    data-css-0=""
    data-css-3=""
    style={
      Object {
        "backgroundColor": "#3598db",
        "bottom": -50,
      }
    }
  >
    <div
      data-css-1=""
      data-css-4=""
    >
      <div
        data-css-5=""
        data-css-2=""
        style={
          Object {
            "width": "100%",
          }
        }
      >
        Hello World
      </div>
    </div>
    <div
      data-css-6=""
    >
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
        data-css-7=""
        data-css-2=""
      />
    </div>
  </div>
</div>
`;
